File: PersianCalendar/src/main/java/com/byagowi/persiancalendar/adapter/TimeItemAdapter.java
package com.byagowi.persiancalendar.adapter;

import android.view.LayoutInflater;
import android.view.ViewGroup;

import com.byagowi.persiancalendar.R;
import com.byagowi.persiancalendar.databinding.TimeItemBinding;
import com.byagowi.persiancalendar.praytimes.Clock;
import com.byagowi.persiancalendar.praytimes.PrayTimes;
import com.byagowi.persiancalendar.util.UIUtils;
import com.google.android.flexbox.FlexboxLayoutManager;

import androidx.annotation.NonNull;
import androidx.annotation.StringRes;
import androidx.recyclerview.widget.RecyclerView;

public class TimeItemAdapter extends RecyclerView.Adapter<TimeItemAdapter.ViewHolder> {
    @StringRes
    static final private int[] timeNames = new int[]{
            R.string.imsak, R.string.fajr, R.string.sunrise, R.string.dhuhr, R.string.asr,
            R.string.sunset, R.string.maghrib, R.string.isha, R.string.midnight
    };
    private PrayTimes mPrayTimes;
    private boolean mExpanded = false;

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        TimeItemBinding binding = TimeItemBinding.inflate(LayoutInflater.from(parent.getContext()),
                parent, false);

        return new ViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        holder.bind(position);
    }

    @Override
    public int getItemCount() {
        return timeNames.length;
    }

    public void setTimes(PrayTimes prayTimes) {
        mPrayTimes = prayTimes;
        for (int i = 0; i < timeNames.length; ++i) notifyItemChanged(i);
    }

    public boolean isExpanded() {
        return mExpanded;
    }

    public void setExpanded(boolean expanded) {
        mExpanded = expanded;
        for (int i = 0; i < timeNames.length; ++i) notifyItemChanged(i);
    }

    class ViewHolder extends RecyclerView.ViewHolder {
        private TimeItemBinding binding;
        private FlexboxLayoutManager.LayoutParams emptyLayout =
                new FlexboxLayoutManager.LayoutParams(0, 0);
        private FlexboxLayoutManager.LayoutParams wrapContent =
                new FlexboxLayoutManager.LayoutParams(
                        FlexboxLayoutManager.LayoutParams.WRAP_CONTENT,
                        FlexboxLayoutManager.LayoutParams.WRAP_CONTENT);

        ViewHolder(TimeItemBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }

        public void bind(int position) {
            int timeName = timeNames[position];
            binding.getRoot().setLayoutParams(!mExpanded && !(timeName == R.string.fajr ||
                    timeName == R.string.dhuhr || timeName == R.string.maghrib)
                    ? emptyLayout : wrapContent);

            binding.name.setText(timeName);

            if (mPrayTimes == null) {
                binding.time.setText("");
                return;
            }

            Clock clock;
            switch (timeName) {
                case R.string.imsak:
                    clock = mPrayTimes.getImsakClock();
                    break;
                case R.string.fajr:
                    clock = mPrayTimes.getFajrClock();
                    break;
                case R.string.sunrise:
                    clock = mPrayTimes.getSunriseClock();
                    break;
                case R.string.dhuhr:
                    clock = mPrayTimes.getDhuhrClock();
                    break;
                case R.string.asr:
                    clock = mPrayTimes.getAsrClock();
                    break;
                case R.string.sunset:
                    clock = mPrayTimes.getSunsetClock();
                    break;
                case R.string.maghrib:
                    clock = mPrayTimes.getMaghribClock();
                    break;
                case R.string.isha:
                    clock = mPrayTimes.getIshaClock();
                    break;
                default:
                case R.string.midnight:
                    clock = mPrayTimes.getMidnightClock();
                    break;
            }
            binding.time.setText(UIUtils.getFormattedClock(clock, false));
        }
    }
}

File: PersianCalendar/src/main/java/com/byagowi/persiancalendar/ui/calendar/times/TimeItemAdapter.kt
package com.byagowi.persiancalendar.ui.calendar.times

import android.view.ViewGroup
import androidx.annotation.StringRes
import androidx.recyclerview.widget.RecyclerView
import com.byagowi.persiancalendar.R
import com.byagowi.persiancalendar.databinding.TimeItemBinding
import com.byagowi.persiancalendar.utils.layoutInflater
import com.byagowi.persiancalendar.utils.toFormattedString
import com.google.android.flexbox.FlexboxLayoutManager
import io.github.persiancalendar.praytimes.PrayTimes

class TimeItemAdapter : RecyclerView.Adapter<TimeItemAdapter.ViewHolder>() {

    @StringRes
    private val timeNames = listOf(
            R.string.imsak, R.string.fajr, R.string.sunrise, R.string.dhuhr, R.string.asr,
            R.string.sunset, R.string.maghrib, R.string.isha, R.string.midnight
    )

    var prayTimes: PrayTimes? = null
        set(prayTimes) {
            field = prayTimes
            timeNames.indices.forEach(::notifyItemChanged)
        }
    var isExpanded = false
        set(expanded) {
            field = expanded
            timeNames.indices.forEach(::notifyItemChanged)
        }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int) = ViewHolder(
            TimeItemBinding.inflate(parent.context.layoutInflater, parent, false)
    )

    override fun onBindViewHolder(holder: ViewHolder, position: Int) = holder.bind(position)

    override fun getItemCount(): Int = timeNames.size

    inner class ViewHolder(private val binding: TimeItemBinding) :
            RecyclerView.ViewHolder(binding.root) {
        private val emptyLayout = FlexboxLayoutManager.LayoutParams(0, 0)
        private val wrapContent = FlexboxLayoutManager.LayoutParams(
                FlexboxLayoutManager.LayoutParams.WRAP_CONTENT,
                FlexboxLayoutManager.LayoutParams.WRAP_CONTENT
        )

        fun bind(position: Int) {
            val timeName = timeNames[position]

            binding.root.layoutParams = if (!isExpanded && timeName !in listOf(
                            R.string.fajr, R.string.dhuhr, R.string.maghrib
                    )
            ) emptyLayout else wrapContent
            binding.name.setText(timeName)
            binding.time.text = prayTimes?.run {
                (when (timeName) {
                    R.string.imsak -> imsakClock
                    R.string.fajr -> fajrClock
                    R.string.sunrise -> sunriseClock
                    R.string.dhuhr -> dhuhrClock
                    R.string.asr -> asrClock
                    R.string.sunset -> sunsetClock
                    R.string.maghrib -> maghribClock
                    R.string.isha -> ishaClock
                    R.string.midnight -> midnightClock
                    else -> midnightClock
                }).toFormattedString()
            } ?: ""
        }
    }
}
